src/Conversion/ONNXToMhlo/NN/Conv.cpp

```diff
@@ -0,0 +1,147 @@
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+
+    ONNXConvOpAdaptor operandAdaptor(operands);
+    ONNXConvOp convOp = llvm::dyn_cast<ONNXConvOp>(op);
+    Location loc = op->getLoc();
+
+    ONNXConvOpShapeHelper shapeHelper(&convOp);
+    LogicalResult shapecomputed = shapeHelper.computeShape(operandAdaptor);
+    assert(succeeded(shapecomputed) && "Could not compute output shape");
+
+    llvm::SmallVector<IndexExpr, 2> kernelShape = shapeHelper.kernelShape;
+    llvm::SmallVector<int64_t, 2> strides = shapeHelper.strides;
+    llvm::SmallVector<int64_t, 2> dilations = shapeHelper.dilations;
+    DimsExpr outputDims = shapeHelper.dimsForOutput();
+    int outputRank = shapeHelper.dimsForOutput().size();
+
+    Value inputOperand = operandAdaptor.X();
+    Value filterOperand = operandAdaptor.W();
+    Value biasOperand = operandAdaptor.B();
+    bool hasBias = !biasOperand.getType().isa<NoneType>();
+    int64_t groupNum = convOp.group();
+
+    assert(isRankedShapedType(inputOperand.getType()) &&
+           "Expected Ranked ShapedType");
+    ShapedType inputType = inputOperand.getType().cast<ShapedType>();
+    llvm::ArrayRef<int64_t> inputShape = inputType.getShape();
+    Type outputType = *op->result_type_begin();
+    // Onnx Input is NCHW
+    int64_t spatialOffset = 2;
+    int64_t rank = inputType.getRank();
+    int64_t kernelSize = kernelShape.size();
+
+    SmallVector<int64_t> inputSpatialDimensions;
+    for (int64_t i = spatialOffset; i < rank; i++) {
+      inputSpatialDimensions.push_back(i);
+    }
+
+    SmallVector<int64_t> kernelDimensions;
+    for (int64_t i = spatialOffset; i < spatialOffset + kernelSize; i++) {
+      kernelDimensions.push_back(i);
+    }
+
+    SmallVector<int64_t> outputSpatialDimensions;
+    for (int64_t i = spatialOffset; i < outputRank; i++) {
+      outputSpatialDimensions.push_back(i);
+    }
+
+    // paddings
+    DimsExpr pads = shapeHelper.pads;
+    llvm::StringRef padding = convOp.auto_pad();
+    int64_t spatialRank = rank - spatialOffset;
+    SmallVector<int64_t> flattenPaddings;
+    bool needPadding = (padding == "NOTSET");
+
+    // currently only support static shape
+    for (int64_t i = 0; i < spatialRank; i++) {
+      if (!needPadding) {
+        if (!pads[i].isLiteral() || !pads[i + spatialRank].isLiteral())
+          return failure();
+      } else {
+        if (!kernelShape[i].isLiteral() || !pads[i].isLiteral() ||
+            !outputDims[i + spatialOffset].isLiteral() ||
+            inputShape[i + spatialOffset] == -1)
+          return failure();
+      }
+    }
+
+    for (int64_t i = 0; i < spatialRank; i++) {
+      if (!needPadding) {
+        flattenPaddings.push_back(pads[i].getLiteral());
+        flattenPaddings.push_back(pads[i + spatialRank].getLiteral());
+      } else {
+        int64_t kdTerm = (kernelShape[i].getLiteral() - 1) * dilations[i] + 1;
```

Comment:
Our shape inference compute left/right top/bottom padding. Is the mhlo padding different than onnx and thus justify recomputing some of it here?

Reply:
Let me help answer this question. `mhlo.convolution` is a generic N-D conv and its padding is of shape (N, 2), where padding[i] specifies lower and upper padding values of spatial dimension i. You can refer to [this](https://www.tensorflow.org/xla/operation_semantics#convwithgeneralpadding_convolution) for more details.

Reply:
Due to ceil mode in ONNX, the padding can be the same even when the output shape differs when auto_pad is NOTSET. Therefore, we need to recalculate the padding. You could refer to @test_default_maxpoolsingleout_strides_nonunifpad and @test_default_maxpoolsingleout_strides_nonunifpad_ceil in both test/mlir/conversion/onnx_to_mhlo/NN/Pooling.mlir and test/mlir/onnx/onnx_shape_inference_maxpool.mlir, and see the difference.